### ISSUE TITLE
Getting a CORRUPT: not on the curve! with elGamal sjcl.ecc.curves.c384

### ISSUE THREAD

**Issue by max-l** [bitwiseshiftleft/sjcl#260]
I have an encryption scheme where I encrypt an array of integers with an
ecc.curves.c384 public key. The private key itself is encrypted in AES, 
and then encrypted by a password.
The idea is that the decrypted private key never goes on the server, only on the client.

This scheme is working "almost" all the time, I occasionally get a situation 
where I encode the integer array with the PubKey, and then get a "CORRUPT: not on the curve!"
when I decode it with the decoded PrivKey.

The problem occurs rarely, aprox every 500 usages.

I have extracted the code that reproduce the same logic (see below) and I ran it 1000s of 
time, and I'm not getting the bug.

It seems that the sjcl library is not deterministic, i.e. that it occasionally gets in a state where
the PubK encodes in a way that the privK cannot decode.

Are there things that I could log when that occurs ?

```
CORRUPT: not on the curve!
q@https:///static/js/lib/sjcl-ecc.js:1:35
fromBits@/static/js/lib/sjcl-ecc.js:79:225
unkem@/static/js/lib/sjcl-ecc.js:87:101
ha@https://static/js/lib/sjcl-ecc.js:50:219
decrypt@https:///static/js/lib/sjcl-ecc.js:51:33
```

```

// version is sjcl#1.0.2
var sjcl = require("./sjcl-ecc.js")



function encryptArrayWithPubKeyThenDecryptWithPrivKeyAndValidate(pubKBase64, arrayToEncode, passwordToDeccryptPrivKey, encryptedPrivateKey) {

  var pubK = new sjcl.ecc.elGamal.publicKey(
    sjcl.ecc.curves.c384,
    sjcl.codec.base64.toBits(pubKBase64)
  )

  var encryptedArray = sjcl.encrypt(pubK, arrayToEncode)

  var sec = new sjcl.ecc.elGamal.secretKey(
    sjcl.ecc.curves.c384,
    sjcl.ecc.curves.c384.field.fromBits(
      sjcl.decrypt(passwordToDeccryptPrivKey, encryptedPrivateKey, {raw: 1})
    )
  )

  // This is where it blows up (CORRUPT: not on the curve!)
  // when it does, haven't got it to occur in a test case...
  var decodedArray = sjcl.decrypt(sec, encryptedArray, {raw: 1})

  if(decodedArray.toString() != arrayToEncode.toString())
    throw new Error("decoded array not the same as input !")
}

function createKeyPairWithEncreyptedPrivateKey(password) {
  var pair = sjcl.ecc.elGamal.generateKeys(sjcl.ecc.curves.c384)
  var pub = pair.pub.get(), sec = pair.sec.get()

  return {
    pubKey: sjcl.codec.base64.fromBits(pub.x.concat(pub.y)),
    encryptedPrivateKey: sjcl.encrypt(password, sec)
  }
}


function randomPassword() {
    var text = "";
    var possible = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
    for( var i=0; i < 10; i++ )
        text += possible.charAt(Math.floor(Math.random() * possible.length));
    return text;
}


for (var k =0 ; k < 1000000; k++) {
  var aPassword = randomPassword()

  var keyPair = createKeyPairWithEncreyptedPrivateKey(aPassword)

  var randomArrayToEncode = sjcl.random.randomWords(256 / 32, 0)

  encryptArrayWithPubKeyThenDecryptWithPrivKeyAndValidate(keyPair.pubKey, randomArrayToEncode, aPassword, keyPair.encryptedPrivateKey)

  if(k % 100 == 1) {
    console.log(k + " cases tested so far")
  }
}

```

**Comment by letorbi** [comment]
Hej max-l,

I ran into a similar problem, when using the k256 ECC curve. I did some research and was finally able to create a pull request (#261), which fixes at least my issue and also the ECC curve k192. The problem was, that the normalize function of the bn class also strips the leading-zeros. This seems to have quite big side-effects so my fix might also help you.

If you don't want to wait for the PR to be merged you can check out [the ecc-curve-fixes branch in my sjcl fork](https://github.com/letorbi/sjcl/tree/ecc-curve-fixes).

Bai
Torben

**Comment by letorbi** [comment]
Hej,

just want to say that my PR has been merged into the main repo.

**Comment by max-l** [comment]
Thanks @letorbi, (just noticed your reply... I need to reenable my github notifications ;-))

**Comment by letorbi** [comment]
Hehe, no problem ;) You should check if my commit really fixes your bug and close this issue, if that's the case.

**Comment by iFA88** [comment]
Greetings, I have today got the same situation with the latest sjcl.js from github. I have crypted my message, and I can not decrpyt them with the same details.
```javascript
function _genKeys(plainPassword) {
	if ( typeof(plainPassword) === '' ) { throw new Error("Password is not initialized!"); }
	var bn = sjcl.bn.fromBits(sjcl.misc.pbkdf2( keccak_256(keccak_256(plainPassword)), sjcl.codec.base64.fromBits(sjcl.codec.hex.toBits(keccak_256(plainPassword).slice(0,32))), 50000, 512));
	bn = bn.mod(sjcl.ecc.curves.c384.r);
	return sjcl.ecc.elGamal.generateKeys(sjcl.ecc.curves.c384, 6, bn);
}
function enCrypt(plainPassword, msg) {
	var keys = this._genKeys(plainPassword);
	return btoa(sjcl.encrypt(new sjcl.ecc.elGamal.publicKey(sjcl.ecc.curves.c384, new sjcl.ecc.point(sjcl.ecc.curves.c384, sjcl.bn.prime.p384.fromBits(keys.pub.get().x), sjcl.bn.prime.p384.fromBits(keys.pub.get().y))), msg));
}
function deCrypt(plainPassword, msg) {
	return sjcl.decrypt(new sjcl.ecc.elGamal.secretKey(sjcl.ecc.curves.c384, sjcl.bn.prime.p384.fromBits(this._genKeys(plainPassword).sec.get())), atob(msg));
}
```
What I have do:
```javascript
 > a = enCrypt("2222", "5ea22449d9981df83a9bb0ee80620b4d2b8e23c9f2a34bfce7b6c0d7beff0113");
eyJpdiI6InZ3SWNNL0lzQXdlVnNLZnBWUmIvYlE9PSIsInYiOjEsIml0ZXIiOjEwMDAwLCJrcyI6MTI4LCJ0cyI6NjQsIm1vZGUiOiJjY20iLCJhZGF0YSI6IiIsImNpcGhlciI6ImFlcyIsImtlbXRhZyI6InFUQzVTamVEN3BCWnh6UTFDa1ZvUGZRVGg0d21lS0pNUnBjc0gxMEptdjJEaHExNnc5cXV6eWJQMEY2a2ZKK1NNN0JYWlY0TG5BZzc0aUVPbVJzamNDaEpGU1EwU2xOV0Z3VFhZcldMQTJzZ1EwNnpJb1BCay9NNXF0UStNMkNSIiwiY3QiOiJrby9DaVFLNjZaTjRxajZJOExiRkw3L3l0eWd6dGppb2R5UXRmNFE5UEdDWFR3ZStMdmhOei9iWSt6cmhWWWFBMkk3RHVzeDFUbFlpVUlBYUJ1VHBiM2RkRVVKNDcyM0EifQ==
 > deCrypt("2222", a);
CORRUPT: not on the curve!
```

This happened only ONCE time, but this should not ever happen.. because this is really sensitive information.

**Comment by iFA88** [comment]
UPDATE:
This happens 8 from 10 on safari browser and Samsung native browser. I think the problem does, that the key size is greater than 256 bit.
